File: typescript/protein-translation/protein-translation.ts
const PROTEIN_TRANSLATIONS: { protein: string; rna: string[] }[] = [
  { protein: 'Methionine', rna: ['AUG'] },
  { protein: 'Phenylalanine', rna: ['UUU', 'UUC'] },
  { protein: 'Leucine', rna: ['UUA', 'UUG'] },
  { protein: 'Serine', rna: ['UCU', 'UCC', 'UCA', 'UCG'] },
  { protein: 'Tyrosine', rna: ['UAU', 'UAC'] },
  { protein: 'Cysteine', rna: ['UGU', 'UGC'] },
  { protein: 'Tryptophan', rna: ['UGG'] },
  { protein: 'STOP', rna: ['UAA', 'UAG', 'UGA'] },
];

const IS_STOP = (protein: string) => protein === 'STOP';

class ProteinTranslation {
  static proteins(rna: string): string[] {
    let result: string[] = [];
    while (rna) {
      let seq: string = rna.substr(0, 3);
      rna = rna.substr(3);

      let translation = PROTEIN_TRANSLATIONS.find((trans) =>
        trans.rna.includes(seq)
      );
      if (translation) {
        if (IS_STOP(translation.protein)) {
          break;
        }
        result.push(translation.protein);
      }
    }

    return result;
  }
}

export default ProteinTranslation;

File: cpp/protein-translation/protein_translation.cpp
#include "protein_translation.h"
#include <stdexcept>

namespace protein_translation
{

    std::vector<std::string> proteins(const std::string &sequence)
    {
        if (sequence.length() % 3 != 0)
        {
            throw std::invalid_argument("Invalid sequence");
        }

        std::vector<std::string> result;

        for (size_t i = 0; i < sequence.length(); i += 3)
        {
            auto v = sequence.substr(i, 3);

            if (v == "AUG")
                result.push_back("Methionine");
            else if (v == "UUU" || v == "UUC")
                result.push_back("Phenylalanine");
            else if (v == "UUA" || v == "UUG")
                result.push_back("Leucine");
            else if (v == "UCU" || v == "UCC" || v == "UCA" || v == "UCG")
                result.push_back("Serine");
            else if (v == "UAU" || v == "UAC")
                result.push_back("Tyrosine");
            else if (v == "UGU" || v == "UGC")
                result.push_back("Cysteine");
            else if (v == "UGG")
                result.push_back("Tryptophan");
            else if (v == "UAA" || v == "UAG" || v == "UGA")
            {
                // result.push_back("STOP");
                break;
            }
            else
                throw std::invalid_argument("Invalid sequence part");
        }

        return result;
    }

} // namespace protein_translation
